pwnlib/encoders/i386/ascii_shellcode.py

```diff
@@ -0,0 +1,300 @@
+            for byte in group(int_size, shellcode, b'\x90'):
+                sc.append(''.join(byte))
+            sc = sc[::-1]
+        # Pack the shellcode to a sub/push sequence
+        for x in map(bytes, sc):
+            for subtraction in self._calc_subtractions(last, x, vocab):
+                if six.PY2:
+                    subtraction = str(subtraction)
+                # sub eax, `subtraction`
+                result += struct.pack('=c{}s'.format(int_size), b'-', subtraction)
+            last = x
+            result += b'P'  # push eax
+        return result
+
+
+    @LocalContext
+    def _calc_subtractions(self, last, target, vocab, max_subs = 4):
+        r""" Given `target` and `last`, return a list of integers that when
+         subtracted from `last` will equal `target` while only constructing
+         integers from bytes in `vocab`
+
+        int_size is take from the context (context.bits / 8)
+
+        Args:
+            last (bytes): Current value of eax
+            target (bytes): Desired value of eax
+            vocab (bytes): Allowed characters
+            max_subs (int): Maximum subtraction attempts
+
+        Raises:
+            ArithmeticError: If a sequence of subtractions could not be found
+
+        Returns:
+            List[bytearray]: List of numbers that would need to be subtracted
+            from `last` to get to `target`
+
+        Examples:
+
+            >>> context.update(arch='i386', os='linux')
+            >>> vocab = b'!"#$%&\'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~'
+            >>> print(encoders.i386.ascii_shellcode.encode._calc_subtractions(b'\x10'*4, b'\x11'*4, vocab))
+            [bytearray(b'!!!!'), bytearray(b'`___'), bytearray(b'~~~~')]
+            >>> print(encoders.i386.ascii_shellcode.encode._calc_subtractions(b'\x11\x12\x13\x14', b'\x15\x16\x17\x18', vocab))
+            [bytearray(b'~}}}'), bytearray(b'~~~~')]
+        """
+        int_size = context.bits // 8
+        subtractions = [bytearray(b'\x00'*int_size)]
+        if six.PY2:
+            last = map(ord, last)
+            target = map(ord, target)
+        for subtraction in range(max_subs):
+            carry = success_count = 0
+            for byte in range(int_size):
+                # Try all combinations of all the characters in vocab of
+                # `subtraction` characters in each combination. So if `max_subs`
+                # is 4 and we're on the second subtraction attempt, products will
+                # equal [\, ", #, %, ...], [\, ", #, %, ...], 0, 0
+                for products in product(
+                    *[x <= subtraction and vocab or (0,) for x in range(max_subs)]
+                ):
+                    if six.PY2:
+                        products = map(lambda x: isinstance(x, str) and ord(x) or x, products)
+                    # Sum up all the products, carry from last byte and the target
+                    attempt = sum(chain(

```

Comment:
I guess it would be more readable to just write
```suggestion
                    attempt = target[byte] + carry + sum(products)
```